File: Code/01/swap.c
#include<stdio.h>

char* s1 =  "hello";
char* s2 =  "world";

void swap(char **x, char **y) {
  char *z = *x;
  *x = *y;
  *y = z;
}

int main(void) {
  swap(&s1, &s2);
  printf("%s\n", s1);
  return 0;
}

File: Code/01/swap.rs

// fn swap(x : &mut String, y : &mut String) {
//   let z : String = *x;
//   *x = *y;
//   *y = z;
// }

fn swap(x : &mut String, y : &mut String) {
  let z : String = x.clone();
  *x = y.clone();
  *y = z;
}

use std::mem;

fn main() {
  let mut s1 : String = String::from("hello");
  let mut s2 : String = String::from("world");
  // mem::swap(&mut s1, &mut s2);
  swap(&mut s1, &mut s2);
  println!("{}", s1);
}
